### DISCUSSION TITLE
How can I add custom routes to a user's app programmatically (plugin-like UX)

### DISCUSSION THREAD

**Discussion by paarthmadan** [remix-run/react-router#11609] [Q&A]
### Context

I'm the developer of a 3rd party app called [Tempo](https://tempolabs.ai) which is responsible for running a user's app (`vite` with `react-router`). We have our own `npm` package that is installed in the user's app.

### Goal

Our app requires the creation of some new routes in the user's app.

Ideally, after installing our package, we should be able to automatically/programatically add routes to the user's app. For example sake, `localhost:5173/tempo-route-a` and `localhost:5173/tempo/route-b`

### More details

Right now, a user adds an initialization hook into their own app when they want to use Tempo:

```tsx
// main.tsx
import React from "react"
import { Tempo } from "tempo-devtools"

Tempo.init() // <------------ Custom initialization hook for our package

ReactDOM.createRoot(document.getElementById("root")!).render(
  <React.StrictMode>
    <BrowserRouter>
        <App />
    </BrowserRouter>
  </React.StrictMode>,
);
```

This decorates their app with helpful artifacts.

### Question

Is there anyway to get access to the list of routes and add more in at runtime? I've looked through the documentation and none of the hooks seem to mutate the state of the routing tree.

I'd also be open to something like:

```tsx
// App.tsx

import { withTempoRoutes } from "tempo-devtools"

function App() {
  return (
    <Routes>
      <Route path="/their-own-route" ...></Route>
      <Route path="/another-route-they-have" ...></Route>
    </Routes>
  )
}

export default withTempoRoutes(App) // <---------- How could I go about implementing this HOC
```

I've played around with creating the HOC shown above, but to no avail. Most of the errors I've encountered have been related to using `Routes` or `useRoutes` outside the context of a `<Router>`. I believe this has to do with referencing `<Routes>` in the package directly.

### Discussion

TL;DR Is there any blessed/recommended path to being able to implement something like this. Ultimately, I want an easy way to add custom routes to a user's app after they run some sort of initializer.
